File: frontend/src/types/extraction.ts
export type ExtractionRunKind = 'materiality' | 'indicators' | 'both'

export type ExtractionRunStatus =
  | 'running'
  | 'completed'
  | 'failed'
  | 'partial'

export type ExtractionTargetKind =
  | 'material_topic'
  | 'sdg_goal'
  | 'indicator_value'

export type ExtractionConfidence = 'high' | 'medium' | 'low'

export type ExtractionSuggestionStatus =
  | 'pending'
  | 'accepted'
  | 'rejected'
  | 'edited'

export type ExtractionProvenance = {
  document_id: string
  document_name: string
  chunk_index: number
  excerpt: string
}

export type MaterialTopicSuggestionPayload = {
  pillar: 'E' | 'S'
  topic: string
  priority: 'alta' | 'media' | 'baixa'
  reasoning?: string
}

export type SdgSuggestionPayload = {
  ods_number: number
  objetivo: string
  acao: string
  indicador: string
  resultado: string
  reasoning?: string
}

export type IndicatorValueSuggestionPayload = {
  template_id: number
  tema: string
  indicador: string
  unidade: string
  value: string
  period?: string | null
  scope?: string | null
  reasoning?: string
}

export type ExtractionSuggestionPayload =
  | MaterialTopicSuggestionPayload
  | SdgSuggestionPayload
  | IndicatorValueSuggestionPayload

export type ExtractionRun = {
  id: string
  project_id: string
  kind: ExtractionRunKind
  status: ExtractionRunStatus
  triggered_by: string | null
  model_used: string | null
  documents_considered: string[] | null
  summary_stats: Record<string, unknown> | null
  error: string | null
  started_at: string
  completed_at: string | null
}

export type ExtractionSuggestion = {
  id: string
  run_id: string
  project_id: string
  target_kind: ExtractionTargetKind
  payload: ExtractionSuggestionPayload
  confidence: ExtractionConfidence
  confidence_score: number | null
  provenance: ExtractionProvenance[]
  conflict_with_existing: boolean
  existing_value_snapshot: Record<string, unknown> | null
  status: ExtractionSuggestionStatus
  reviewed_at: string | null
  reviewed_by: string | null
  reviewer_notes: string | null
  created_at: string
}

export type ExtractionSuggestionList = {
  items: ExtractionSuggestion[]
  total: number
}

export type ExtractionStreamEventType =
  | 'run_started'
  | 'extractor_started'
  | 'suggestion'
  | 'extractor_completed'
  | 'run_completed'
  | 'error'

export type ExtractionStreamEvent =
  | { type: 'run_started'; data: { run_id: string; kind: ExtractionRunKind; model: string | null } }
  | {
      type: 'extractor_started'
      data: { kind: 'materiality' | 'indicators' }
    }
  | { type: 'suggestion'; data: ExtractionSuggestion }
  | {
      type: 'extractor_completed'
      data: { kind: 'materiality' | 'indicators'; count: number }
    }
  | {
      type: 'run_completed'
      data: { run_id: string; status: ExtractionRunStatus; summary: Record<string, unknown> | null }
    }
  | { type: 'error'; data: { message: string; extractor?: string } }

export type UpdateExtractionSuggestionInput = {
  action: 'accept' | 'reject' | 'edit'
  payload?: ExtractionSuggestionPayload
  notes?: string
}

export type BulkExtractionSuggestionInput = {
  ids: string[]
  action: 'accept_all' | 'reject_all'
}

export type BulkExtractionSuggestionResponse = {
  succeeded: string[]
  failed: { id: string; detail: string }[]
}

File: backend/app/schemas/extraction.py
from datetime import datetime
from decimal import Decimal
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import (
    ExtractionConfidence,
    ExtractionRunKind,
    ExtractionRunStatus,
    ExtractionSuggestionStatus,
    ExtractionTargetKind,
)

# ---------------------------------------------------------------------------
# LLM-internal models — used with ChatOpenAI.with_structured_output(...)
# ---------------------------------------------------------------------------


class ProvenanceItem(BaseModel):
    """Pointer back to the document chunk that supports a suggestion."""

    document_id: UUID = Field(description="ID of the source Document row.")
    document_name: str = Field(description="Original filename for display.")
    chunk_index: int = Field(description="Index of the chunk inside the document.")
    excerpt: str = Field(
        description="Short verbatim quote (≤400 chars) supporting the extraction.",
        max_length=400,
    )


class MaterialTopicSuggestion(BaseModel):
    """A material topic extracted from the documents."""

    pillar: Literal["E", "S"] = Field(description="E for environmental, S for social.")
    topic: str = Field(
        description=(
            "GRI disclosure code in the format 'GRI X-Y' (e.g. 'GRI 302-1'). "
            "Must be a real GRI code present in the standards reference."
        ),
        min_length=1,
        max_length=64,
    )
    priority: Literal["alta", "media", "baixa"] = Field(
        description="Priority based on signal strength found in the documents."
    )
    confidence: ExtractionConfidence
    reasoning: str = Field(
        description="One sentence explaining why this topic was extracted.",
        max_length=500,
    )
    provenance: list[ProvenanceItem] = Field(min_length=1)


class SdgSuggestion(BaseModel):
    """A priority SDG (Sustainable Development Goal) extracted from documents."""

    ods_number: int = Field(ge=1, le=17)
    objetivo: str = Field(
        description="Short title of the SDG (Portuguese).",
        min_length=1,
        max_length=255,
    )
    acao: str = Field(
        default="",
        description="Concrete action found in the documents (optional).",
        max_length=2000,
    )
    indicador: str = Field(
        default="",
        description="Indicator that tracks this SDG action (optional).",
        max_length=2000,
    )
    resultado: str = Field(
        default="",
        description="Reported result/outcome (optional).",
        max_length=2000,
    )
    confidence: ExtractionConfidence
    reasoning: str = Field(max_length=500)
    provenance: list[ProvenanceItem] = Field(min_length=1)


class IndicatorValueSuggestion(BaseModel):
    """A numeric value extracted to fill an IndicatorTemplate row."""

    template_id: int = Field(
        description="Primary key of the IndicatorTemplate this value belongs to."
    )
    tema: str = Field(min_length=1, max_length=64)
    indicador: str = Field(min_length=1, max_length=255)
    unidade: str = Field(
        description="Unit found in the source. Should match the template's unidade.",
        max_length=64,
    )
    value: str = Field(
        description=(
            "Value as found in the source, normalized to a plain number string "
            "(no thousands separators, dot as decimal). Empty string if no usable "
            "value was found — in that case do NOT emit this suggestion."
        ),
        min_length=1,
        max_length=255,
    )
    period: str | None = Field(
        default=None,
        description="Reporting period (e.g. '2024', 'FY2024'). Optional.",
        max_length=64,
    )
    scope: str | None = Field(
        default=None,
        description="Optional scope qualifier (e.g. 'Escopo 1' for emissions).",
        max_length=64,
    )
    confidence: ExtractionConfidence
    reasoning: str = Field(max_length=500)
    provenance: list[ProvenanceItem] = Field(min_length=1)


class MaterialityExtraction(BaseModel):
    """Top-level structured output for the materiality extractor."""

    material_topics: list[MaterialTopicSuggestion] = Field(default_factory=list)
    sdg_goals: list[SdgSuggestion] = Field(default_factory=list)


class IndicatorsExtraction(BaseModel):
    """Top-level structured output for the indicators extractor."""

    values: list[IndicatorValueSuggestion] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# API schemas — request/response
# ---------------------------------------------------------------------------


class StartExtractionRequest(BaseModel):
    kind: ExtractionRunKind = ExtractionRunKind.BOTH


class ExtractionRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    project_id: UUID
    kind: ExtractionRunKind
    status: ExtractionRunStatus
    triggered_by: UUID | None
    model_used: str | None
    documents_considered: list[Any] | None
    summary_stats: dict[str, Any] | None
    error: str | None
    started_at: datetime
    completed_at: datetime | None


class ExtractionSuggestionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    run_id: UUID
    project_id: UUID
    target_kind: ExtractionTargetKind
    payload: dict[str, Any]
    confidence: ExtractionConfidence
    confidence_score: Decimal | None
    provenance: list[Any]
    conflict_with_existing: bool
    existing_value_snapshot: dict[str, Any] | None
    status: ExtractionSuggestionStatus
    reviewed_at: datetime | None
    reviewed_by: UUID | None
    reviewer_notes: str | None
    created_at: datetime


class UpdateSuggestionRequest(BaseModel):
    action: Literal["accept", "reject", "edit"]
    payload: dict[str, Any] | None = None
    notes: str | None = None


class BulkUpdateRequest(BaseModel):
    ids: list[UUID] = Field(min_length=1)
    action: Literal["accept_all", "reject_all"]


class ExtractionSuggestionList(BaseModel):
    items: list[ExtractionSuggestionResponse]
    total: int
